### ISSUE TITLE
Performance against `jax.experimental.ode.odeint`

### ISSUE THREAD

**Issue by Jordan-Dennis** [patrick-kidger/diffrax#76]
Hi @patrick-kidger,
I was excited to test out `Diffrax` in our code. However, we found it did not perform as well as expected. This is likely to nuances on our end, but because o https://github.com/google/jax/issues/9654 I thought I would post a MWE. 

```py
import diffrax
import jax
import ticktack

PARAMS = (774.86, 0.25, 0.8, 6.44)

STEADY_PROD = 1.8803862513018528

STEADY_STATE = jax.numpy.array(
    [1.34432991e+02, 7.07000000e+02, 1.18701144e+03,
    3.95666872e+00, 4.49574232e+04, 1.55056740e+02,
    6.32017337e+02, 4.22182768e+02, 1.80125397e+03,
    6.63307283e+02, 7.28080320e+03], 
    dtype=jax.numpy.float64)

PROD_COEFFS = jax.numpy.array(
    [0.7, 0.3, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], 
    dtype=jax.numpy.float64)

MATRIX = jax.numpy.array([
    [-0.509, 0.009, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
    [0.508, -0.44, 0.068, 0.0, 0.0, 0.545, 0.0, 0.167, 0.002, 0.002, 0.0],
    [0.0, 0.121, -0.155, 12.0, 0.001, 0.0, 0.0, 0.003, 0.0, 0.0, 0.0],
    [0.0, 0.0, 4.4000e-02, -1.3333e+01, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
    [0.0, 0.0, 0.042, 1.333, -0.001, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
    [0.0, 0.229, 0.0, 0.0, 0.0, -1.046, 0.0, 0.0, 0.0, 0.0, 0.0],
    [0.0, 0.0, 0.0, 0.0, 0.0, 0.136, -0.033, 0.0, 0.0, 0.0, 0.0],
    [0.0, 0.0, 0.0, 0.0, 0.0, 0.364, 0.033, -0.183, 0.0, 0.0, 0.0],
    [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.01, -0.002, 0.0, 0.0],
    [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.003, 0.0, -0.002, 0.0],
    [0.0, 0.0, 3.333e-04, 0.0, 5.291e-06, 0.0, 0.0, 0.0, 0.0, 4.0e-04, -1.2340e-04]], 
    dtype=jax.numpy.float64)

@jax.jit 
def driving_term(t, args):
    start_time, duration, phase, area = jax.numpy.array(args)
    middle = start_time + duration / 2.
    height = area / duration

    gauss = height * \
        jax.numpy.exp(- ((t - middle) / (0.5 * duration)) ** 16.)
    sine = STEADY_PROD + 0.18 * STEADY_PROD *\
        jax.numpy.sin(2 * jax.numpy.pi / 11 * t + phase * 2 * jax.numpy.pi / 11)

    return (sine + gauss) * 3.747

@jax.jit
def jax_dydt(y, t, args, /, matrix=MATRIX, production=driving_term, 
                   prod_coeffs=PROD_COEFFS):
    ans = jax.numpy.matmul(matrix, y)
    production_rate_constant = production(t, args)
    production_term = prod_coeffs * production_rate_constant
    return ans + production_term

@jax.jit
def diffrax_dydt(t, y, args, /, matrix=MATRIX, production=driving_term, 
                 prod_coeffs=PROD_COEFFS):
    ans = jax.numpy.matmul(matrix, y)
    production_rate_constant = production(t, args)
    production_term = prod_coeffs * production_rate_constant
    return ans + production_term

time_out = jax.numpy.linspace(750, 800, 1000)

%%timeit
jax.experimental.ode.odeint(jax_dydt, STEADY_STATE, time_out, PARAMS)

term = diffrax.ODETerm(diffrax_dydt)
solver = diffrax.Bosh3()
step_size = diffrax.PIDController(rtol=1e-10, atol=1e-10)
save_time = diffrax.SaveAt(ts=time_out)

%%timeit
diffrax.diffeqsolve(args=PARAMS, terms=term, solver=solver, y0=STEADY_STATE,
                    t0=time_out.min(), t1=time_out.max(), dt0=0.01,
                    saveat=save_time, stepsize_controller=step_size, 
                    max_steps=10000)
```

Sorry that the example is so volumous but I wanted to keep it very similar to our code.

Thanks in advance.

Jordan

**Comment by patrick-kidger** [comment]
Hey! So the good news is that all that is going on here is that you've passed the wrong arguments into `diffeqsolve` / are timing the wrong things. Once these are fixed then Diffrax comes out in the lead!

As for what's going on here:

1. You've specified very different solver tolerances. `jax.experimental.ode.odeint` has default tolerances of `rtol=1.4e-8`, `atol=1.4e-8`. (Which you have not overridden.) Meanwhile you've specified much-more-stringent tolerances of `1e-10` for Diffrax.
2. Next, your choice of solver. `jax.experimental.ode.odeint` uses Dormand--Prince 5(4), which is `diffrax.Dopri5`. This is usually much more efficient than `Bosh3`, which is what you're using with Diffrax at the moment.
3. You're timing compile time + run time, not just run time. The first time you call either `jax.experimental.ode.odeint` or `diffrax.diffeqsolve`, then you hit the JIT compiler. The correct thing to do is to run it once to let it compile, then time the second invocation.
4. As a very minor point, `jax.experimental.ode.odeint` figures out the initial step size automatically, whilst with Diffrax you've specified that it must be `dt0=0.01`. In this case if you set `dt0=None` then Diffrax will also figure out the initial step size automatically, and moreover this does seem to improve things over your default.

In terms of what results I get when the above things are fixed:
- `jax.experimental.ode.odeint` takes 300 steps in 1815400 nanoseconds.
- Diffrax, with the faulty choices you've made above, take 6124 in 2037835300 nanoseconds
- Diffrax, with the fixes I've suggested, take 296 steps in 1159300 nanoseconds. That is to say, Diffrax is ~57% faster!

**Comment by patrick-kidger** [comment]
By the way, I saw you mention taking gradients over in https://github.com/google/jax/issues/9654. To head off some common queries about this: by default `jax.experimental.ode.odeint` and Diffrax use different algorithms for backpropagating through a diffeq solve. Diffrax's is more accurate, but may be faster/slower depending on the problem.

If you use the default backpropagation method in Diffrax, then you can minimise the time spent backpropagating by reducing `max_steps` to the smallest value that works for your problem. The time spent scales as `O(log max_steps)`. To be precise, it increases every time `max_steps` passes a power of 16.

If you want to switch to the alternative algorithm used by `jax.experimental.ode.odeint` -- which will produce less accurate results -- then you can do this by passing `adjoint=diffrax.BacksolveAdjoint()` to `diffrax.diffeqsolve`. As I say, this can be faster or slower depending on your problem.

**Comment by Jordan-Dennis** [comment]
Thanks heaps.
I'll retest this and post the results in https://github.com/google/jax/issues/9654. Funny you should mention the different solvers. We initially used `DifferentialEquations.jl` to profile different solver performances for our specific problem. `BS3()` just won out over `DP5()` but had a major gradient advantage. Obviously this is pretty sensitive to implementation specifics though. I'll also test this difference and share the result in this thread.
Thanks again 
Jordan

**Comment by Jordan-Dennis** [comment]
As you predicted the `Dopri5()` solver was faster. Thanks again!

**Comment by Jordan-Dennis** [comment]
Hi @patrick-kidger,
I've continued to work with `diffrax` and when reintegrating it into the source code for our package ran into the same (but smaller) performance hit, with both the gradients and base call. I have tested this pretty thoroughly (I think) and believe that I have located the source of the issue. It seems that calling `diffrax.diffeqsolve` from within a `jit` decorated function is slower than the `jax.experimental.ode.odeint` called from within a `jit` decorated function. I've put together MWE (again quite long, but much condensed compared to the package code). 
```py
import diffrax
import jax
from jax.experimental import ode
jax.config.update("jax_enable_x64", True)

PARAMS = jax.numpy.array([1.0, 1.0], dtype=jax.numpy.float64)
MATRIX = jax.numpy.array([[2.0, 1.0], [1.0, 2.0]], dtype=jax.numpy.float64)
Y_INIT = jax.numpy.array([0.0, 0.0], dtype=jax.numpy.float64)

@jax.jit
def driving_term(t, args):
    return jax.numpy.array([jax.numpy.exp(t), jax.numpy.sin(t)]) * args

@jax.jit
def jax_ode(y, t, args):
    return MATRIX @ y + driving_term(t, args)

@diffrax.ODETerm
@jax.jit
def diffrax_ode(t, y, args):
    return MATRIX @ y + driving_term(t, args)

@jax.jit
def y(t):
    y_1 = 0.05 * jax.numpy.exp(3 * t) - 0.25 * jax.numpy.exp(t) + \
        0.2 * jax.numpy.cos(t) + 0.1 * jax.numpy.sin(t)
    y_2 = 0.05 * jax.numpy.exp(3 * t) + 0.25 * jax.numpy.exp(t) + \
        - 0.3 * jax.numpy.cos(t) + 0.4 * jax.numpy.sin(t)
    return jax.numpy.array([y_1, y_2])

time = jax.numpy.linspace(0, 1, 100)
data = y(time) + jax.random.normal(jax.random.PRNGKey(0), (2, 100))

@jax.jit
def jax_loss(args, /, dydx=jax_ode, y0=Y_INIT, t=time, data=data.T):
    solution = ode.odeint(dydx, y0, t, args)
    sum_squares = (solution - data) ** 2
    return jax.numpy.sum(sum_squares)

%%timeit
jax_loss(PARAMS)

step_size = diffrax.PIDController(rtol=1.4e-8, atol=1.4e-8)
times = diffrax.SaveAt(ts=time)
t_initial = time.min()
t_final = time.max()

adjoint_controller = diffrax.PIDController(
    norm=diffrax.adjoint_rms_seminorm, 
    rtol=1.4e-8, atol=1.4e-8)

grad = diffrax.BacksolveAdjoint(
    stepsize_controller=adjoint_controller,
    solver=diffrax.Dopri5())

solver = diffrax.Dopri5()

@jax.jit
def diffrax_loss(args, /, t0=t_initial, t1=t_final, dydx=diffrax_ode, 
    adjoint=grad, steps=step_size, saves=times, y0=Y_INIT, solver=solver, 
    data=data.T):
    solution = diffrax.diffeqsolve(
        t0=t0, t1=t1, dt0=None, terms=dydx, stepsize_controller=steps, 
        saveat=saves, adjoint=adjoint, solver=solver, args=args, y0=y0)
    sum_squares = (solution.ys - data) ** 2
    return jax.numpy.sum(sum_squares)

%%timeit
diffrax_loss(PARAMS)

jax_grad = jax.jit(jax.grad(jax_loss))
diffrax_grad = jax.jit(jax.grad(diffrax_loss))

%%timeit
jax_grad(PARAMS)

%%timeit
diffrax_grad(PARAMS)
```
The `%%timeit` cells can be run consecutively to get the post `jit` times. 
Thanks again and sorry for another long example
Regards
Jordan

**Comment by patrick-kidger** [comment]
You're still timing both compile time and runtime together.

**Comment by patrick-kidger** [comment]
Having dug into this a bit.

A) First of all, just for your own interest, here's how the MWE can be simplified whilst retaining essentially the same performance characteristics.

```python
import time
import diffrax
import jax
from jax.experimental import ode
jax.config.update("jax_enable_x64", True)

PARAMS = jax.numpy.array([1.0, 1.0])
Y_INIT = jax.numpy.array([0.0, 0.0])
TIMES = jax.numpy.linspace(0, 1, 100)

def myode(t, y, args):
    return jax.numpy.stack([jax.numpy.exp(t), jax.numpy.sin(t)]) * args

@jax.jit
def jax_loss(args):
    _myode = lambda y, t, args: myode(t, y, args)
    solution = ode.odeint(_myode, Y_INIT, TIMES, args)
    return jax.numpy.sum(solution)

@jax.jit
def diffrax_loss(args):
    solution = diffrax.diffeqsolve(
        terms=diffrax.ODETerm(myode),
        solver=diffrax.Dopri5(),
        t0=TIMES.min(),
        t1=TIMES.max(),
        dt0=None,
        y0=Y_INIT,
        args=args,
        stepsize_controller=diffrax.PIDController(rtol=1.4e-8, atol=1.4e-8),
        saveat=diffrax.SaveAt(ts=TIMES),
        adjoint=diffrax.BacksolveAdjoint(),
    )
    return jax.numpy.sum(solution.ys)

jax_grad = jax.jit(jax.grad(jax_loss))
diffrax_grad = jax.jit(jax.grad(diffrax_loss))

def timeit(name, fn):
    fn(PARAMS)
    start = time.time()
    out = fn(PARAMS)
    end = time.time()
    print(name, out, end - start)

timeit("jax fwd", jax_loss)
timeit("diffrax fwd", diffrax_loss)
timeit("jax bwd", jax_grad)
timeit("diffrax bwd", diffrax_grad)
```

In particular you can see that I've removed a lot of extraneous stuff that didn't need to be there. (`data`, `y(t)`,  several `jax.jit`s, separate `jax_ode` and `diffrax_ode`, etc.)

Running this I get
```
jax fwd 87.89542131286058 1.8596649169921875e-05
diffrax fwd 87.89542019476409 0.0001049041748046875
jax bwd [71.97048838 15.92492968] 0.00010991096496582031
diffrax bwd [71.97048829 15.92492966] 0.0019135475158691406
```

B) Next up, I'm afraid this isn't a great example for demonstrating performance problems. It's completely true that on this example, Diffrax is slower by roughly 10^-3 seconds... but that's not very much. This kind of difference in speed could be because of all sorts of pretty small overheads.

Whilst it would be great to optimise Diffrax that tiny amount further, I don't think trying to shave off 10^-3 seconds is ever going to be very high on the priority list!

[
Side note: for very small microbenchmarks like this it's generally better to measure the difference between two competing libraries / etc. additivity ("10^-3 seconds") -- as I'm doing -- rather than multiplicatively ("10 times slower"). There are some general principles I can dig into here if you're curious.
]

C) Okay, how can we make this a good benchmark? Well, basically: make it bigger! We need to get timings that are more than just tiny fractions of a second. We can do this in the above example by replacing `TIMES` with
```
TIMES = jax.numpy.linspace(0, 1, 10000)
```
and running the code now produces
```
jax fwd 8768.320954233159 0.00013589859008789062
diffrax fwd 8768.32086312595 0.0003516674041748047
jax bwd [7182.959158   1585.36147877] 0.005400419235229492
diffrax bwd [7182.959158   1585.36147877] 0.1751558780670166
```

Oh no -- looking at `diffrax bwd`, that seems like a huge slowdown when using Diffrax!

The good news is that I've figured out where this slowdown is coming from. Whenever you run `diffeqsolve` then Diffrax performs a number of checks to make sure that the user-supplied `dt0 > 0`, that the integration hasn't hit one of its failure conditions (e.g. hitting the maximum number of steps) etc. It turns out that some of this was erroneously happening at the Python level rather than the XLA level, and as a result couldn't be optimised away by the XLA compiler.

This will be closed in #77. After these fixes are in then the times drop substantially to:
```
jax fwd 8768.320954233159 0.00014257431030273438
diffrax fwd 8768.32086312595 0.00023984909057617188
jax bwd [7182.959158   1585.36147877] 0.0051496028900146484
diffrax bwd [7182.959158   1585.36147877] 0.0043659210205078125
```
And in particular the slowdown on `diffrax bwd` has gone.

**Comment by Jordan-Dennis** [comment]
Thanks for the help. I'll integrate these changes into our code :smile:.
